tests/Feature/MailExportTest.php

```diff
@@ -0,0 +1,176 @@
+            'filename' => false,
+        ]);
+
+        Storage::fake('local');
+
+        Mail::send(new StandardMailable);
+
+        $this->assertEmpty(Storage::disk('local')->files('mail-export'));
+    }
+
+    /** @test */
+    public function it_will_export_mails_as_eml_files()
+    {
+        config()->set('mail-export', [
+            'enabled' => true,
+            'disk' => 'local',
+            'path' => 'mail-export',
+            'filename' => 'some_filename',
+        ]);
+
+        Storage::fake('local');
+
+        Mail::send(new ExportableMailable);
+
+        Storage::disk('local')->assertExists('mail-export/some_filename.eml');
+    }
+
+    /** @test */
+    public function it_can_export_using_the_default_application_disk_if_none_is_given()
+    {
+        config()->set('filesystem.default', 'local');
+        config()->set('mail-export', [
+            'enabled' => true,
+            'disk' => null,
+            'path' => 'mail-export',
+            'filename' => false,
+        ]);
+
+        Storage::fake('local');
+        Storage::fake('some_disk');
+
+        Mail::send(new ExportableMailable);
+
+        $this->assertCount(1, Storage::disk('local')->files('mail-export'));
+        $this->assertEmpty(Storage::disk('some_disk')->files('mail-export'));
+    }
+
+    /** @test */
+    public function it_can_export_using_storage_details_configured_by_mailable_property()
+    {
+        config()->set('mail-export', [
+            'enabled' => true,
+            'disk' => 'local',
+            'path' => 'mail-export',
+            'filename' => false,
+        ]);
+
+        Storage::fake('local');
+        Storage::fake('some_disk');
+
+        Mail::send(new ExportableByPropertiesMailable);
+
+        Storage::disk('some_disk')->assertExists('some_path/some_filename.eml');
+        $this->assertEmpty(Storage::disk('local')->files('mail-export'));
+    }
+
+    /** @test */
+    public function it_can_export_using_storage_details_configured_by_mailable_method()
+    {
+        config()->set('mail-export', [
+            'enabled' => true,
+            'disk' => 'local',
+            'path' => 'mail-export',
+            'filename' => false,
+        ]);
+
+        Storage::fake('local');
+        Storage::fake('some_disk');
+
+        Mail::send(new ExportableByMethodsMailable);
+
+        Storage::disk('some_disk')->assertExists('some_path/some_filename.eml');
+        $this->assertEmpty(Storage::disk('local')->files('mail-export'));
+    }
+
+    /** @test */
+    public function it_can_export_using_an_autogenerated_filename_if_none_is_given()
+    {
+        config()->set('mail-export', [
+            'enabled' => true,
+            'disk' => 'local',
+            'path' => 'mail-export',
+            'filename' => false,
+        ]);
+
+        Storage::fake('local');
+        Carbon::setTestNow(Carbon::create(2021, 3, 26, 15, 01, 42));
+
+        Mail::send(new ExportableMailable);
```

Comment:
it would be nice if we could inject this in the constructor to that the assertion makes more sense.

Reply:
sorry I don't get it...

Reply:
We are asserting against something that's hardcoded in the class and outside the context of this test, so whilst we have the right assertion, it's not immediately obvious to the person reading the test where the values come from, without clicking through to the `ExpotableMailable` class.

What would be nice:

```php
Mail::send(new ExportableMailable(['to' => 'jane@example.com', 'subject' => 'this is the subject']));

Storage::disk('local')
            ->assertExists('mail-export/2021_03_26_150142_jane_at_example_com_this_is_the_subject.eml');
```

That way, we know right away the values we are expecting to see in the output as we controlled the input. 

But can be one for later tbh - just UX - not functional.

Reply:
Aye sure makes sense 👍